rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,568 @@
+- `ApplicationHooks`: Provides hooks for the application lifecycle, including startup, shutdown, and error handling.
+- `AuthenticationHooks`: Provides hooks for authentication and authorization logic.
+- `TelemetryHooks`: Provides hooks for OpenTelemetry spans and metrics.
+- `OperationHooks`: Provides hooks for parsed, normalized, and planned GraphQL operations.
+- `Module`: Provides hooks for the module lifecycle, including provisioning and shutdown.
+
+```go
+type MyModule struct{}
+
+type Subgraph struct {
+	// The name of the subgraph
+	Name string
+	// The ID of the subgraph
+	ID string
+	// The URL of the subgraph
+	URL *url.URL
+	// The schema of the subgraph
+	Schema *graphql.Schema
+}
+
+type Graph struct {
+	// The name of the graph
+	Name string
+	// The ID of the graph
+	ID string
+	// Version of the graph
+	Version string
+	// Subgraphs of the graph
+	Subgraphs []*core.Subgraph
+	// GraphQL schema of the graph
+	Schema *graphql.Schema
+}
+
+type GatewayRequest struct {
+	// The original GraphQL request with all the information like query, variables, operation name, extensions etc.
+	Request *core.GraphQLRequest
+	// The parsed, normaliazed and planned operation with all the information like name, variables, type, document representation,
+	// client name version, uploaded files, plan, normalization, persisted operation etc.
+	Operation *core.Operation
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The active graph
+	Graph *core.Graph
+	// The original HTTP request
+	Orignal *http.Request
+	// Logger for the gateway request
+	Logger *zap.Logger
+}
+
+type GatewayResponse struct {
+	// The original RouterRequest
+	Request *core.GatewayRequest
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The final GraphQL response with all the information like data, errors, extensions etc.
+	// This is the response that will be sent to the client and can be manipulated or replaced
+	Response *core.GraphQLResponse
+	// The original HTTP response
+	Orignal *http.Response
+}
+
+type GatewayHooks interface {
+	// OnRequest is called when a request is made to the gateway and after all GraphQL information is available
+	// Returning an error will result in a GraphQL error being returned to the client.
+	OnGatewayRequest(ctx *core.GatewayRequest, err error) error
+	// OnResponse is called before the response is sent to the client
+	// Returning an error will result in a GraphQL error being returned to the client.
+	OnGatewayResponse(ctx *core.GatewayResponse, err error) error
+	// OnError is called when an error occurs during the gateway lifecycle
+	OnGatewayError(err error)
+}
+
+type SubgraphRequest struct {
+	// The active subgraph
+	Subgraph *core.Subgraph
+	// The original Gateway request
+	GatewayRequest *core.RouterRequest
+	// The final GraphQL request to the subgraph
+	Request *core.GraphQLRequest
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
```

Comment:
What is `core.Telemetry` and what funcs does it expose?
Is this to create spans and metrics?
Does it contain the context to properly propagate spans?